core/src/main/kotlin/org/sert2521/sertain/events/EventBus.kt

```diff
@@ -19,15 +19,27 @@ suspend fun <E : Event> fire(event: E) {
 
 @UseExperimental(kotlinx.coroutines.ExperimentalCoroutinesApi::class, kotlinx.coroutines.FlowPreview::class)
 inline fun <reified E : Event> CoroutineScope.subscribe(noinline action: suspend (E) -> Unit) =
-        launch {
-            events.asFlow().filter { it is E }.map { it as E }.apply { collect(action) }
-        }
+            launch {
+                events.asFlow().filter { it is E }.map { it as E }.apply {
+                    collect {
+                        launch {
+                            action(it)
+                        }
```

Comment:
Did you ever figure out why this works?

Reply:
Well, I think that `collect` doesn't launch a new coroutine, and so the code running just stays running. Using `launch` though starts a new coroutine, and so it gets canceled when the other coroutines get canceled.